c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+//! Add(move _2, const 1)` stores the result of `a + 1` into a local.  If needed,
+//! we could extend the `unlower` pass to also record that `move _2` (a.k.a. `bb0[2]`
+//! `[Rvalue, RvalueOperand(0)]`) is lowered from the `Expr` `a`.
+//!
+//! On `g`, the unlowering map includes the following (among other entries):
+//!
+//! ```text
+//! bb0[5]: Terminator { source_info: ..., kind: _4 = f(move _5) -> [return: bb1, unwind: bb2] }
+//!   []: StoreIntoLocal, `f(y)`
+//!   [Rvalue]: Expr, `f(y)`
+//!   [Rvalue, CallArg(0)]: Expr, `y`
+//! bb1[1]: _0 = Add(move _3, move _4)
+//!   []: StoreIntoLocal, `x + f(y)`
+//!   [Rvalue]: Expr, `x + f(y)`
+//! ```
+//!
+//! The call terminator `_4 = f(move _5)` computes `f(y)` and stores the result
+//! into a local; its rvalue is `f(y)` itself, and the first argument of the rvalue
+//! is `y`.
+
+use crate::panic_detail;
+use crate::rewrite::build_span_index;
+use crate::rewrite::expr::mir_op::SubLoc;
+use crate::rewrite::span_index::SpanIndex;
+use log::*;
+use rustc_hir as hir;
+use rustc_hir::intravisit;
+use rustc_hir::HirId;
+use rustc_middle::hir::nested_filter;
+use rustc_middle::mir::{self, Body, Location};
+use rustc_middle::ty::{DefIdTree, TyCtxt};
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+use std::collections::btree_map::{BTreeMap, Entry};
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub struct MirOrigin {
+    pub hir_id: HirId,
+    pub span: Span,
+    pub desc: MirOriginDesc,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub enum MirOriginDesc {
+    /// This MIR represents the whole HIR expression.
+    Expr,
+    /// This MIR stores the result of the HIR expression into a MIR local of some kind.
+    StoreIntoLocal,
+}
+
+struct UnlowerVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    mir: &'a Body<'tcx>,
+    span_index: SpanIndex<Location>,
+    /// Maps MIR (sub)locations to the HIR node that produced each one, if known.
+    mir_map: BTreeMap<(Location, Vec<SubLoc>), MirOrigin>,
+}
+
+impl<'a, 'tcx> UnlowerVisitor<'a, 'tcx> {
+    fn location_span(&self, loc: Location) -> Span {
+        self.mir
+            .stmt_at(loc)
+            .either(|stmt| stmt.source_info.span, |term| term.source_info.span)
+    }
+
+    fn record(&mut self, loc: Location, sub_loc: &[SubLoc], ex: &hir::Expr) {
+        self.record_desc(loc, sub_loc, ex, MirOriginDesc::Expr);
+    }
+
+    fn record_desc(
+        &mut self,
+        loc: Location,
+        sub_loc: &[SubLoc],
+        ex: &hir::Expr,
+        desc: MirOriginDesc,
+    ) {
+        let origin = MirOrigin {
+            hir_id: ex.hir_id,
+            span: ex.span,
+            desc,
+        };
+        match self.mir_map.entry((loc, sub_loc.to_owned())) {
+            Entry::Vacant(e) => {
+                e.insert(origin);
+            }
+            Entry::Occupied(e) => {
+                if *e.get() != origin {
+                    let old_origin = e.get().clone();
```

Comment:
```suggestion
                    let old_origin = *e.get();
```

Reply:
I applied the other suggestion instead, which supersedes this